internal/webhooks/v1beta1/metal3cluster_webhook.go

```diff
@@ -0,0 +1,160 @@
+	c, ok := obj.(*infrav1.Metal3Cluster)
+	if !ok {
+		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected a Metal3Cluster but got a %T", obj))
+	}
+
+	return nil, webhook.validate(nil, c)
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type.
+func (webhook *Metal3Cluster) ValidateUpdate(_ context.Context, oldObj, newObj runtime.Object) (admission.Warnings, error) {
+	newM3C, ok := newObj.(*infrav1.Metal3Cluster)
+	if !ok {
+		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected a Metal3Cluster but got a %T", newObj))
+	}
+
+	oldM3C, ok := oldObj.(*infrav1.Metal3Cluster)
+	if !ok || oldM3C == nil {
+		return nil, apierrors.NewBadRequest(fmt.Sprintf("expected a Metal3Cluster but got a %T", oldObj))
+	}
+
+	return nil, webhook.validate(oldM3C, newM3C)
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type.
+func (webhook *Metal3Cluster) ValidateDelete(_ context.Context, _ runtime.Object) (admission.Warnings, error) {
+	return nil, nil
+}
+
+func (webhook *Metal3Cluster) validate(oldM3C, newM3C *infrav1.Metal3Cluster) error {
+	var allErrs field.ErrorList
+	if newM3C.Spec.ControlPlaneEndpoint.Host == "" {
+		allErrs = append(
+			allErrs,
+			field.Invalid(
+				field.NewPath("spec", "controlPlaneEndpoint"),
+				newM3C.Spec.ControlPlaneEndpoint.Host,
+				"is required",
+			),
+		)
+	}
+
+	if newM3C.Spec.CloudProviderEnabled != nil && newM3C.Spec.NoCloudProvider != nil {
+		if *newM3C.Spec.CloudProviderEnabled == *newM3C.Spec.NoCloudProvider {
+			allErrs = append(
+				allErrs,
+				field.Invalid(
+					field.NewPath("spec", "cloudProviderEnabled"),
+					newM3C.Spec.CloudProviderEnabled,
+					"cloudProviderEnabled conflicts the value of noCloudProvider",
+				),
+			)
+		}
+	}
+
+	if oldM3C != nil {
+		// Validate cloudProviderEnabled
+		if newM3C.Spec.CloudProviderEnabled != nil && oldM3C.Spec.NoCloudProvider != nil {
+			if *newM3C.Spec.CloudProviderEnabled == *oldM3C.Spec.NoCloudProvider {
+				allErrs = append(
+					allErrs,
+					field.Invalid(
+						field.NewPath("spec", "cloudProviderEnabled"),
+						newM3C.Spec.CloudProviderEnabled,
+						"ValidateUpdate failed, cloudProviderEnabled conflicts the value of noCloudProvider",
+					),
+				)
+			}
+		}
+
+		// Validate noCloudProvider
+		if newM3C.Spec.NoCloudProvider != nil && oldM3C.Spec.CloudProviderEnabled != nil {
+			if *newM3C.Spec.NoCloudProvider == *oldM3C.Spec.CloudProviderEnabled {
+				allErrs = append(
+					allErrs,
+					field.Invalid(
+						field.NewPath("spec", "noCloudProvider"),
+						newM3C.Spec.NoCloudProvider,
+						"ValidateUpdate failed, noCloudProvider conflicts the value of cloudProviderEnabled",
+					),
+				)
+			}
+		}
+	}
+
+	if len(allErrs) == 0 {
+		return nil
+	}
+	return apierrors.NewInvalid(infrav1.GroupVersion.WithKind("Metal3Cluster").GroupKind(), newM3C.Name, allErrs)
+}
```

Comment:
So for this release cycle we will have two webhook definitions for the same object? How would it work in practicality ? i.e. validating webhooks would be run from which function definition from here or from `api/` dir? Also I see now the webhook definitions in `api/` any kubebuilder verbs

Reply:
I got the answer while reviewing the `main.go` that we are mainly using the `internal/` webhook definition. I am now thinking is there anypoint leaving the `api/` webhook code just for deprecation? Perhaps CAPI is doing it that way?

Reply:
If someone is importing the CAPM3 API and using any of the webhooks functions in their code, removing them immediately could break their code. While I don’t expect anyone to be using them, to be cautious, let’s deprecate them first before removal.